### PR TITLE
use cobra and upstream command machinery to run kube-rbac-proxy

### PR DESCRIPTION
**Pull request by stlaz**
This PR makes use of the cobra package for setting up the command. This is to bring the codebase closer to what upstream is doing.

I also split the large main file into several packages so that it matches upstream better, and so it might be easier to do the move towards the generic server options later.

/assign @ibihim

### CODE REVIEW

File: cmd/app/kube-rbac-proxy.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+		Args: func(cmd *cobra.Command, args []string) error {
+			for _, arg := range args {
+				if len(arg) > 0 {
+					return fmt.Errorf("%q does not take any arguments, got %q", cmd.CommandPath(), args)
+				}
+			}
+			return nil
+		},
+	}
+	fs := cmd.Flags()
```
Comment (ibihim) [inline cmd/app/kube-rbac-proxy.go:237]: Nit: Newline above. Plz

**Review by ibihim — COMMENTED** [review]
I am not a fan of Cobra, but it looks way better and the separation of "Complete" and "Validate" is awesome.

Thank you a lot!

File: cmd/app/kube-rbac-proxy_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	"github.com/brancz/kube-rbac-proxy/pkg/authz"
+	"github.com/google/go-cmp/cmp"
+)
+
+func Test_parseAuthorizationConfigFile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply cmd/app/kube-rbac-proxy_test.go:29]: this is the generated unit tests' way to get around testing private functions

File: cmd/app/kube-rbac-proxy.go [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+		allowPaths:  o.AllowPaths,
+		ignorePaths: o.IgnorePaths,
+	}
+
+	completed.upstreamURL, err = url.Parse(o.Upstream)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse upstream URL: %w", err)
+	}
+
+	if upstreamCAPath := o.UpstreamCAFile; len(upstreamCAPath) > 0 {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply cmd/app/kube-rbac-proxy.go:158]: I like to use scoped variables for condition blocks

File: cmd/app/kube-rbac-proxy.go [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+
+For more information, please go to https://github.com/brancz/kube-rbac-proxy/issues/196
+
+===============================================
+		`)
+
+	}
+
+	if len(o.allowPaths) > 0 && len(o.ignorePaths) > 0 {
+		klog.Fatal("Cannot use --allow-paths and --ignore-paths together.")
```
[the remark this answers is not in the document]
Comment (stlaz) [reply cmd/app/kube-rbac-proxy.go:122]: that must be a copy-paste mistake

**Pull request merged**